### PR TITLE
v.dissolve: Allow all types of features to be dissolved

### PR DESCRIPTION
**Pull request by ecodiv**
This seems to solve issue #3420

### CODE REVIEW

**Comment by wenzeslaus** [comment]
Would you be able to provide a test for this? The current tests pass so that's great. A test for lines or points would be nice addition. v.dissolve has grass.gunittest-based tests in `testsuite` directory and pytest-based tests in `tests`. You can pick whatever makes more sense to you.

**Comment by ecodiv** [comment]
> Would you be able to provide a test for this? The current tests pass so that's great. A test for lines or points would be nice addition. v.dissolve has grass.gunittest-based tests in `testsuite` directory and pytest-based tests in `tests`. You can pick whatever makes more sense to you.

I haven't figured out how those tests work. Something on my todo list.

**Review by wenzeslaus — DISMISSED** [review]
`v.extract -d` operates on boundaries, but the dissolve for the other types happens in v.reclass if I understand that correctly, so this looks good. (For the record, the areas limitation was there at least since the translation from Bash to Python for v7.)

There is a lot of tests for the current functionality including some new ones through v.clip, so the tests would be nice, but perhaps not crucial.

There is, however, several places in the code and doc which mention area. Can you please update those?

**Comment by ecodiv** [comment]
> There is, however, several places in the code and doc which mention area. Can you please update those?

I can give it a try. Most descriptions you read are about dissolving boundaries between adjacent polygons. That is the more common and also easily visualized user case. But in this case, adjacent or overlapping features can be dissolved irrespective of their type (at least, I tried with lines and polygons, I haven't tried the other types). That makes it a bit harder to clearly describe.

But perhaps first the observation that in the manual, it reads that:

> Multiple areas with the same category or the same attribute value which are not adjacent are merged together into one entity which consists of multiple areas, i.e., a multipolygon.

This is different behavior from the general description, which reads:

> Dissolves boundaries between adjacent areas sharing a common category number or attribute. 

I would have thought that the latter is the more common approach to dissolving features? 

I actually like how QGIS has implemented this. Its dissolve function offers the option to "keep disjoint features as separate". Selecting that option means that any features in the layer that do not share a common boundary with another feature will be kept as separate features.  I am not sure how ArcGIS handles things.

**Comment by wenzeslaus** [comment]
> > There is, however, several places in the code and doc which mention area. Can you please update those?
> 
> I can give it a try. Most descriptions you read are about dissolving boundaries between adjacent polygons. That is the more common and also easily visualized user case.

For merging this PR, it would be enough just to fix the places where is says it does areas (only). If it says "areas and other types" that should be enough to clarify it is not areas only. 

> But perhaps first the observation that in the manual, it reads that:
> 
> > Multiple areas with the same category or the same attribute value which are not adjacent are merged together into one entity which consists of multiple areas, i.e., a multipolygon.
> 
> This is the actual behavior, which differs from what the general description suggests, which reads:
> 
> > Dissolves boundaries between adjacent areas sharing a common category number or attribute.
> 
> I would have thought that the latter is the more common approach to dissolving features?

Based on what I found, just a conjecture based on what are the defaults in different SW, the more common behavior is the same what the current v.dissolve does. It may be different from the expectations created by the more geometrically focused examples in GIS 101 classes, but that's the reality perhaps based on what is easier to implement. The v.dissolve implementation relies on two-step (three with aggregation) process with the geometry and attributes being resolved separately and at this point does not have an easy switch for changing to the other behavior. The update would have to happen in v.extract and practically it is likely just a question of how much time/money is someone willing to spend adding that feature.

**Comment by ecodiv** [comment]
> > > There is, however, several places in the code and doc which mention area. Can you please update those?
> > 
> > 
> > I can give it a try. Most descriptions you read are about dissolving boundaries between adjacent polygons. That is the more common and also easily visualized user case.
> 
> For merging this PR, it would be enough just to fix the places where is says it does areas (only). If it says "areas and other types" that should be enough to clarify it is not areas only.
> 
> > But perhaps first the observation that in the manual, it reads that:
> > > Multiple areas with the same category or the same attribute value which are not adjacent are merged together into one entity which consists of multiple areas, i.e., a multipolygon.
> > 
> > 
> > This is the actual behavior, which differs from what the general description suggests, which reads:
> > > Dissolves boundaries between adjacent areas sharing a common category number or attribute.
> > 
> > 
> > I would have thought that the latter is the more common approach to dissolving features?
> 
> Based on what I found, just a conjecture based on what are the defaults in different SW, the more common behavior is the same what the current v.dissolve does. It may be different from the expectations created by the more geometrically focused examples in GIS 101 classes, but that's the reality perhaps based on what is easier to implement. The v.dissolve implementation relies on two-step (three with aggregation) process with the geometry and attributes being resolved separately and at this point does not have an easy switch for changing to the other behavior. The update would have to happen in v.extract and practically it is likely just a question of how much time/money is someone willing to spend adding that feature.

Fair enough. And it is possible to separate geographically separated features in a separate step using the code https://github.com/OSGeo/grass-addons/pull/1021, of that is required.

As for the text, I'll have a look.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
Are you sure about the image the result has A and B values in key column, there should probably be cat column again with 1 and 2.

**Comment by ecodiv** [comment]
> Are you sure about the image the result has A and B values in key column, there should probably be cat column again with 1 and 2.

You are right, I misunderstood the earlier text. Removed that part of the text and the figure.

File: scripts/v.dissolve/v.dissolve.html [hunk 1]
```diff
@@ -1,126 +1,176 @@
...
 </center>
 
-<h3>Merging behavior</h3>
+<p>
+
+Instead of dissolving features based on the category values, the user 
+can define an integer or string column using the <b>column</b> 
+parameter. In that case, features that share the same value in that 
+column are dissolved. Note, the newly created layer does not retain the 
+cat values from the input layer, only the values of the new key column.
```
Comment (wenzeslaus) [inline scripts/v.dissolve/v.dissolve.html:22]: It "does retain values of new key column"? That does not make sense to me. Is this related to your comment about the figure?

File: scripts/v.dissolve/v.dissolve.html [hunk 2]
```diff
@@ -1,126 +1,176 @@
...
 
-Multiple areas with the same category or the same attribute value
-which are not adjacent are merged together into one entity
-which consists of multiple areas, i.e., a multipolygon.
+<p>
+Note that multiple areas with the same category or the same attribute 
+value that are not adjacent are merged into one entity, which consists 
+of multiple features. I.e., this results in multipart features.
```
Comment (wenzeslaus) [inline scripts/v.dissolve/v.dissolve.html:27]:
```suggestion
Note that multiple areas with the same category or the same attribute 
value that are not adjacent are merged into one entity, which consists 
of multiple features, i.e., a multipart feature.
```

Or simply start a new sentence and restate the same in other words (Or say "In other words, ...").

File: scripts/v.dissolve/v.dissolve.html [hunk 3]
```diff
@@ -1,126 +1,176 @@
...
+future.
 
 <p>
-If only <b>aggregate_columns</b> is provided, methods default to
-<em>n</em>, <em>min</em>, <em>max</em>, <em>mean</em>, and <em>sum</em>.
-If the <em>univar</em> backend is specified, all the available methods
-for the <em>univar</em> backend are used.
+If the <b>result_columns</b> is not provided, each method is applied to 
+each column specified by <b>aggregate_columns</b>. This results in 
+columns for all combinations. These result columns have auto-generated
```
Comment (wenzeslaus) [inline scripts/v.dissolve/v.dissolve.html:107]:
```suggestion
each column specified by <b>aggregate_columns</b>. This results in a column for each of the combinations. These result columns have auto-generated 
```

File: scripts/v.dissolve/v.dissolve.html [hunk 4]
```diff
@@ -1,126 +1,176 @@
...
+names based on the aggregate column and method. For example, setting 
+the following parameters:
+
+<div class="code"><pre>
+aggregate_columns=A,B
+aggregate_methods=sum,n
+</pre></div>
+
+<p>
+Results in the following columns: A_sum, A_n, B_sum, B_n. Also see
```
Comment (wenzeslaus) [inline scripts/v.dissolve/v.dissolve.html:117]: Is this common to start here with a capital letter when it is continuation of a sentence from above? You could also split the "For example..." part and the result part into two sentences to avoid this completely.

Same issue couple lines above.

File: scripts/v.dissolve/v.dissolve.html [hunk 5]
```diff
@@ -1,126 +1,176 @@
...
+the following parameters:
+
+<div class="code"><pre>
+aggregate_columns=A,B
+aggregate_methods=sum,n
+</pre></div>
+
+<p>
+Results in the following columns: A_sum, A_n, B_sum, B_n. Also see 
+the Example section.
```
Comment (wenzeslaus) [inline scripts/v.dissolve/v.dissolve.html:118]: Or Examples section? Same below.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
I don't know if it was the formatting changes or just GitHub diff rendering, but I had to read everything to figure out what was changed. That took some time to get to it. It looks great overall. Hopefully next round can be quick.

**Comment by ecodiv** [comment]
> I don't know if it was the formatting changes or just GitHub diff rendering, but I had to read everything to figure out what was changed. That took some time to get to it. It looks great overall. Hopefully next round can be quick.

Sorry about that. I have never figured out when GitHub shows the diff for the line in which the change took place, or whole blocks of text, or what I should do to make sure it only shows the line with changes.

File: scripts/v.dissolve/v.dissolve.html [hunk 6]
```diff
@@ -1,126 +1,176 @@
...
-which <em>v.db.univar</em> uses by default,
-i.e., <em>n</em>, <em>min</em>, <em>max</em>, <em>range</em>,
-<em>mean</em>, <em>mean_abs</em>, <em>variance</em>, <em>stddev</em>,
-<em>coef_var</em>, and <em>sum</em>.
-When the <em>sql</em> backend is used, the methods in turn depends on the SQL
-database backend used for the attribute table of the input vector.
-For SQLite, it is at least the following
-<a href="https://www.sqlite.org/lang_aggfunc.html">build-in aggregate functions</a>:
-<em>count</em>, <em>min</em>, <em>max</em>,
+The default behavior is typically intended for interactive use and
```
[the remark this answers is not in the document]
Comment (ecodiv) [reply scripts/v.dissolve/v.dissolve.html:41]: I am not sure why I used the word typically here.

File: scripts/v.dissolve/v.dissolve.html [hunk 7]
```diff
@@ -1,126 +1,176 @@
...
+
+<h4>aggregation using univar</h4>
+
+When <em>univar</em> is used, the methods available are the ones which 
+<em>v.db.univar</em> uses by default, i.e., <em>n</em>, <em>min</em>, 
+<em>max</em>, <em>range</em>, <em>mean</em>, <em>mean_abs</em>, 
+<em>variance</em>, <em>stddev</em>, <em>coef_var</em>, and 
+<em>sum</em>.
+
+<h4>Aggregation using univar</h4>
```
[the remark this answers is not in the document]
Comment (ecodiv) [reply scripts/v.dissolve/v.dissolve.html:56]: Corrected

File: scripts/v.dissolve/v.dissolve.html [hunk 8]
```diff
@@ -1,126 +1,176 @@
...
+functions applied, e.g.,
 
 <p>
-Aggregate methods are specified by name in <b>aggregate_methods</b>
-or using SQL syntax in <b>aggregate_columns</b>.
-If <b>result_columns</b> is provided including type information
-and the <em>sql</em> backend is used,
-<b>aggregate_columns</b> can contain SQL syntax specifying both columns
-and the functions applied, e.g.,
 <code>aggregate_columns="sum(cows) / sum(animals)"</code>.
```
[the remark this answers is not in the document]
Comment (ecodiv) [reply scripts/v.dissolve/v.dissolve.html:88]: Should there be a <p> above  a code block?

File: scripts/v.dissolve/v.dissolve.html (hunk 1 above)
Comment (ecodiv) [reply scripts/v.dissolve/v.dissolve.html]: I think. I was a bit confused about that part in the original text.  The important part, I guess, from the original text was that the original cat values are not retained. Deleting the second part of the sentence.

File: scripts/v.dissolve/v.dissolve.html [hunk 9]
```diff
@@ -1,126 +1,177 @@
...
-<a href="https://www.sqlite.org/lang_aggfunc.html">build-in aggregate functions</a>:
-<em>count</em>, <em>min</em>, <em>max</em>,
+The default behavior is intended for interactive use and 
+testing. For scripting and other automated usage, explicitly specifying 
+the backend with the <b>aggregate_backend</b> parameter is strongly 
+recommended. When choosing, note that the <em>sql</em> aggregate 
+backend, regardless of the underlying database, will typically perform 
+significantly better than the <em>univar</em> backend.
+
+<h4>aggregation using univar</h4>
```
Comment (wenzeslaus) [inline scripts/v.dissolve/v.dissolve.html:48]:
```suggestion
<h4>Aggregation using univar backend</h4>
```

File: scripts/v.dissolve/v.dissolve.html [hunk 10]
```diff
@@ -1,126 +1,177 @@
...
+
+When <em>univar</em> is used, the methods available are the ones which 
+<em>v.db.univar</em> uses by default, i.e., <em>n</em>, <em>min</em>, 
+<em>max</em>, <em>range</em>, <em>mean</em>, <em>mean_abs</em>, 
+<em>variance</em>, <em>stddev</em>, <em>coef_var</em>, and 
+<em>sum</em>.
+
+<h4>Aggregation using sql</h4>
```
Comment (wenzeslaus) [inline scripts/v.dissolve/v.dissolve.html:56]:
```suggestion
<h4>Aggregation using sql backend</h4>
```

**Review by wenzeslaus — APPROVED** [review]

**Review by wenzeslaus — APPROVED** [review]

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
Actually, the size of the images changed slightly, but nothing is changed otherwise. That needs to be reverted. (I will do it.)

**Review by wenzeslaus — APPROVED** [review]

**Comment by wenzeslaus** [comment]
> ...I have never figured out when GitHub shows the diff for the line in which the change took place, or whole blocks of text, or what I should do to make sure it only shows the line with changes.

If there are changes in wrapping of paragraphs, than it shows the whole blocks as edited. Both Git and GitHub can show just changed words and they try to show changes around unchanged text, but if there is a change on every line, than everything looks new and the word diff is no longer possible. The versioning is line-oriented and if there is a change in a line it will show it (and it will not show it otherwise).

You can limit some of these issue by really editing just the necessary stuff, e.g., by picking only what you intended to change using `git add -p` (which is a really great tool in general). This may help with things like slightly modified images or automatically changed lines by your editor.

If you do changes everywhere, then, of course, you create a huge diff and that's fine if making changes everywhere is what you are really doing.
